### ISSUE TITLE
List Zip Node

### ISSUE THREAD

**Issue by LeonCheung** [JacquesLucke/animation_nodes#370]
One usage of it is for element sorting, since a ready-to-use sorting tool hasn't been implemented yet. current workaround can be like this:

![sorting](https://cloud.githubusercontent.com/assets/494788/12451882/ae27c952-bfc6-11e5-9225-b5c65453a4c6.jpg)

Of course there can be other usefulnesses in different cases, so a possible node for that can be nice and easy to use (or another better solution?)

**Comment by og76** [comment]
take a look here: 
https://github.com/og76/animation_nodes/tree/working-tests/nodes/geometry
and see sort vectors, sort objects

Well, not sure how updated they are, *I was just workin on this*, for example on vectors I was just doin a radial sort etc  ..

But I'll propose a separate sorting issue as that has some interesting cases

**Comment by og76** [comment]
sorry, 
about the very question:
the zip you do inside the expression, why do you need it outside?

    [a for (a,b,c) in sorted( zip(listA, listB, listC), key = lambda e: e[1] - e[2] ) , reversed = rev]

where listA, listB, listC are separate inputs

**Comment by LeonCheung** [comment]
Oh I did use it inside a single expression in one of the BA post, here just want to expose the "pain", even by nesting into one expression, especially for artists who know little about scripting.

![sortss](https://cloud.githubusercontent.com/assets/494788/12454524/d5ebee9e-bfd2-11e5-9516-0807236823c8.jpg)

**Comment by og76** [comment]
I can see that (and the pain) but if they know little about scripting, chances are they know not of zip list.
And they probably fight with loops ..

However, is there another use? Loops are already zipping lists

**Comment by LeonCheung** [comment]
Hmm, another case would be zipping multiple lists into tuple, to make lists for vector, edge, polygon or custom prop etc.

**Comment by og76** [comment]
here too, should the length of result list be the *shortest, longest or a specific one?*

also, am I going to get *a tuple list* out of it or can I get only *a for (abc)* in ABc.. ?

**Comment by LeonCheung** [comment]
Depends. Omg I have to sleep now, sorry. night guys... :zzz:

**Comment by og76** [comment]
good night :)

**Comment by LeonCheung** [comment]
I think a basic Zip node should only require lists of same length, and output as generic / generic list. It seems to be the last basic operator that AN should have.

However, for lists of different length, maybe like below if without the Zip node:

![fill](https://cloud.githubusercontent.com/assets/494788/14242843/beb373b8-fa84-11e5-9e6b-8d7c08ec9f11.png)

Basically Zip can be very possible to be used together with Sort node, imho.

Also, as we mentioned above, it is useful to create tuple things for some cases.

**Comment by JacquesLucke** [comment]
The problem I have with a zip node is that it outputs a generic list. And I'd like to have as less generic sockets as possible..  
What we can have is a node that fills a list up to a specified length using different methodes (use last; repeat; fallback value; ..)

If you want to sort with by two properties you can also use two sort nodes. Eg if you have a list of objects and want to sort them by Z coordinate. But when two objects have the same Z value they should be sorted by X. You archieve that by sorting by X first and then sort the sorted list again by Z. This works because python uses a stable sorting algorithm.

**Comment by LeonCheung** [comment]
you mean like this?

![sort2](https://cloud.githubusercontent.com/assets/494788/14244298/ef4c1284-fa8c-11e5-82f3-770c25fb56d0.png)

But it should never work like this. Or I missed some point here?

**Comment by JacquesLucke** [comment]
yes. exactly like this. Have you tested it? (I didn't but it should work)

**Comment by JacquesLucke** [comment]
https://en.wikipedia.org/wiki/Sorting_algorithm#Stability

**Comment by LeonCheung** [comment]
As seen in the screenshot, the result was no expected. I think the second sort node treats the previous one as an unsoertred litst.

**Comment by JacquesLucke** [comment]
Yes. They treat the list as if it is unsorted, but the sort function doesn't change the order of objects that have the same key. You can play with the reverse sockets in this file:
http://www.pasteall.org/blend/41310

![multisort](https://cloud.githubusercontent.com/assets/5881326/14246451/75398294-fa67-11e5-9562-d26ed3e0c13b.PNG)

**Comment by LeonCheung** [comment]
I can see the point, just still puzzling on the case I played with:
![sort3](https://cloud.githubusercontent.com/assets/494788/14248946/f24248ec-faa9-11e5-88ec-bc0c27a432d1.png)

According to the rule you mentioned, shoudn't the final result be ["Ä","D","B","C"]?
http://www.pasteall.org/blend/41312

**Comment by og76** [comment]
this is the precision issue 
(messing with your supposed stable sorting that would allow cascading)
as I said somewhere else, you'd be surprised .. in a bad way

(compare and you'll see)

**Comment by JacquesLucke** [comment]
the issue here has nothing to do with precision. Precision only is a problem when you do different calculations which in theory should result in the same solution but don't. When two values are exactly the same, because you set them to be the same they always compare equal.

@LeonCheung: the problem you have here is that the sort nodes are in the wrong order. The "main sorting" should be the last. and the "secondary sort" before that.
Because when multiple values compare equal in the main sort they will stay in the same order the secondary sort put them in... not sure if that was a good explanation.... 

Bottom line is that if you change the order or the sort nodes you get the result you looked for.

**Comment by LeonCheung** [comment]
Ah, OK, i seems so, I'll try learn more avbout it, Many thanks! Both. ;)

**Comment by og76** [comment]
I guess all this Zip node is going down to: **list fill** + bonus **list repeat** 
as in #483 and respective pulls mentioned there. soon to be here ..

**Comment by LeonCheung** [comment]
Hmm, after quite a few weeks, I still think a ready-to-use Zip node is much more convenient for working with the new Sort List node, especially for complex sorting like this:

![zip](https://cloud.githubusercontent.com/assets/494788/15804604/5e3c9828-2b41-11e6-87c8-999691279b61.png)

Any further idea?

**Comment by JacquesLucke** [comment]
Maybe I still don't understand your goal, but I think you can archieve it in this way:
![multisort2](https://cloud.githubusercontent.com/assets/5881326/15804646/8ccc4282-2b10-11e6-9a38-7d0a0d80acb3.PNG)

**Comment by LeonCheung** [comment]
Hmm, luckily it works fine in my particular case. Thanks. Will see if Zip node is really needed in other cases, will see. :)

**Comment by LeonCheung** [comment]
Close for now before finding a convincing cases.
